### PR TITLE
8276337: Use override specifier in HeapDumper

### PR DESCRIPTION
**Pull request by lkorinth**
Use override specifier in HeapDumper. It is safer. 

/label add serviceability

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276337](https://bugs.openjdk.java.net/browse/JDK-8276337): Use override specifier in HeapDumper

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6274/head:pull/6274` \
`$ git checkout pull/6274`

Update a local copy of the PR: \
`$ git checkout pull/6274` \
`$ git pull https://git.openjdk.java.net/jdk pull/6274/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6274`

View PR using the GUI difftool: \
`$ git pr show -t 6274`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6274.diff">https://git.openjdk.java.net/jdk/pull/6274.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
In what way is it "safer" ?? What are you trying to guard against?

**Comment by lkorinth** [comment]
> In what way is it "safer" ?? What are you trying to guard against?

If you use `override`, you can not forget to update a method signature if you change the method signature in the base class. You can also not by mistake change the method signature in this class without realizing you should change the base class. There is no downside of using `override` that I know of.

**Comment by lkorinth** [comment]
I have some new code in the pipe, and I want to use `override` in that code without changing unrelated code in heapDumper.cpp. Thus I first make this change so that my future change can keep the "style" of the file.

File: src/hotspot/share/services/heapDumper.cpp [hunk 1]
```diff
@@ -607,23 +607,23 @@ class DumpWriter : public AbstractDumpWriter {
...
+  char const* error() const override    { return _backend.error(); }
 
   // Called by threads used for parallel writing.
   void writer_loop()                    { _backend.thread_loop(); }
   // Called when finish to release the threads.
-  virtual void deactivate()             { flush(); _backend.deactivate(); }
+  void deactivate() override            { flush(); _backend.deactivate(); }
   // Get the backend pointer, used by parallel dump writer.
-  CompressionBackend* backend_ptr() { return &_backend; }
+  CompressionBackend* backend_ptr()     { return &_backend; }
```
Comment (tstuefe) [inline src/hotspot/share/services/heapDumper.cpp:626]: Unrelated, but could be a const method, no?

**Review by tstuefe — APPROVED** [review]
Hi Leo, seems fine.

..Thomas

**Comment by tstuefe** [comment]
> > In what way is it "safer" ?? What are you trying to guard against?
> 
> If you use `override`, you can not forget to update a method signature if you change the method signature in the base class. You can also not by mistake change the method signature in this class without realizing you should change the base class. There is no downside of using `override` that I know of.

In addition to that, its a helpful documentation for the code reader.

File: src/hotspot/share/services/heapDumper.cpp (hunk 1 above)
Comment (lkorinth) [reply src/hotspot/share/services/heapDumper.cpp:626]: Not easily I think, if I make the method const, I would have to return a const pointer, and then I need to do const operations on the backend etc.

**Comment by dholmes-ora** [comment]
The only "downside" to using `override` is that if you use it on one method then you must use it on all applicable methods in a class, so it can tend to have a viral affect in terms of the scope of changes needed just to add the case(s) you're really interested in.

My experience with `override` in Java is a negative one as I often fail to add the annotation and then get chastised by the compiler for my oversight. It might be useful when writing larger libraries with class hierarchies but I find it a hinderence when writing simple code like tests.

For C++ I find the placement of override awkward and less visible than virtual.

**Review by dholmes-ora — APPROVED** [review]

**Comment by lkorinth** [comment]
Thanks Thomas and David!
/integrate

**Pull request closed without merging**